consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.ratis.statemachine.SnapshotInfo;
+import org.apache.ratis.statemachine.SnapshotRetentionPolicy;
+import org.apache.ratis.statemachine.StateMachineStorage;
+import org.apache.ratis.statemachine.impl.FileListSnapshotInfo;
+import org.apache.ratis.util.MD5FileUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * TODO: Warning, currently in Ratis 2.2.0, there is a bug in installSnapshot. In subsequent
+ * installSnapshot, a follower may fail to install while the leader assume it success. This bug will
+ * be triggered when the snapshot threshold is low. This is fixed in current Ratis Master, and
+ * hopefully will be introduced in Ratis 2.3.0.
+ */
+public class SnapshotStorage implements StateMachineStorage {
+  private IStateMachine applicationStateMachine;
+  private File stateMachineDir;
+  private Logger logger = LoggerFactory.getLogger(SnapshotStorage.class);
+
+  public SnapshotStorage(IStateMachine applicationStateMachine) {
+    this.applicationStateMachine = applicationStateMachine;
+  }
+
+  private ByteBuffer getMetadataFromTermIndex(TermIndex termIndex) {
+    String ordinal = String.format("%d_%d", termIndex.getTerm(), termIndex.getIndex());
+    ByteBuffer metadata = ByteBuffer.wrap(ordinal.getBytes());
+    return metadata;
+  }
+
+  private TermIndex getTermIndexFromMetadata(ByteBuffer metadata) {
+    Charset charset = Charset.defaultCharset();
+    CharBuffer charBuffer = charset.decode(metadata);
+    String ordinal = charBuffer.toString();
+    String[] items = ordinal.split("_");
+    return TermIndex.valueOf(Long.parseLong(items[0]), Long.parseLong(items[1]));
+  }
+
+  @Override
+  public void init(RaftStorage raftStorage) throws IOException {
+    this.stateMachineDir = raftStorage.getStorageDir().getStateMachineDir();
+  }
+
+  @Override
+  public SnapshotInfo getLatestSnapshot() {
+    SnapshotMeta snapshotMeta = applicationStateMachine.getLatestSnapshot(stateMachineDir);
+    if (snapshotMeta == null) {
+      return null;
+    }
+    TermIndex snapshotTermIndex = getTermIndexFromMetadata(snapshotMeta.getMetadata());
+
+    List<FileInfo> fileInfos = new ArrayList<>();
+    try {
+      for (File file : snapshotMeta.getSnapshotFiles()) {
+        Path filePath = file.toPath();
+        MD5Hash fileHash = MD5FileUtil.computeMd5ForFile(file);
+        FileInfo fileInfo = new FileInfo(filePath, fileHash);
+        fileInfos.add(fileInfo);
+      }
+    } catch (IOException ioException) {
+      logger.error("read file info failed for snapshot file ", ioException);
+    }
+
+    return new FileListSnapshotInfo(
+        fileInfos, snapshotTermIndex.getTerm(), snapshotTermIndex.getIndex());
+  }
+
+  public void takeSnapshot(TermIndex lastApplied) {
```

Comment:
Why we put this method in `StateMachineStorage`, seems weird

Reply:
done